pkg/xen-tools/patches-4.14.0/13-qemu-Revamp-mmap-allocation-routine.patch

```diff
@@ -0,0 +1,124 @@
+From 3ad5ce46a00385fc59adb5ba9f7c4b358e346bbf Mon Sep 17 00:00:00 2001
+From: Sergey Temerkhanov <s.temerkhanov@gmail.com>
+Date: Sun, 20 Sep 2020 02:44:45 +0300
+Subject: [PATCH 2/2] qemu: Revamp mmap allocation routine
+
+Do not double-map the memory addresses
+
+Signed-off-by: Sergey Temerkhanov <s.temerkhanov@gmail.com>
+---
+ util/mmap-alloc.c | 42 +++++++++++++++---------------------------
+ 1 file changed, 15 insertions(+), 27 deletions(-)
+
+diff --git a/util/mmap-alloc.c b/util/mmap-alloc.c
+index 27dcccd8ec..7ff2883e42 100644
+--- a/tools/qemu-xen/util/mmap-alloc.c
++++ b/tools/qemu-xen/util/mmap-alloc.c
+@@ -89,19 +89,18 @@ void *qemu_ram_mmap(int fd,
+                     bool is_pmem)
+ {
+     int flags;
++    int mapfd;
+     int map_sync_flags = 0;
+-    int guardfd;
+     size_t offset;
+     size_t pagesize;
+     size_t total;
+-    void *guardptr;
+     void *ptr;
++    int ret;
+ 
+     /*
+      * Note: this always allocates at least one extra page of virtual address
+      * space, even if size is already aligned.
+      */
+-    total = size + align;
+ 
+ #if defined(__powerpc64__) && defined(__linux__)
+     /* On ppc64 mappings in the same segment (aka slice) must share the same
+@@ -115,39 +114,31 @@ void *qemu_ram_mmap(int fd,
+     flags = MAP_PRIVATE;
+     pagesize = qemu_fd_getpagesize(fd);
+     if (fd == -1 || pagesize == qemu_real_host_page_size) {
+-        guardfd = -1;
++        mapfd = -1;
+         flags |= MAP_ANONYMOUS;
+     } else {
+-        guardfd = fd;
++        mapfd = fd;
+         flags |= MAP_NORESERVE;
+     }
+ #else
+-    guardfd = -1;
++    mapfd = -1;
+     pagesize = qemu_real_host_page_size;
+     flags = MAP_PRIVATE | MAP_ANONYMOUS;
+ #endif
+ 
+-    guardptr = mmap(0, total, PROT_NONE, flags, guardfd, 0);
+-
```

Comment:
OK, so here we're basically mapping the entire region requested as PROT_NONE -- effectively "pre-allocating" it only to do `mmap(guardptr + offset, size, PROT_READ | PROT_WRITE, flags | map_sync_flags, fd, 0);` which, if my understanding is correct, will effectively provide us with a guard page at the beggining and everything else starting from offset is going to get re-mapped `PROT_READ | PROT_WRITE`.

Now, of course, Linux may actually refuse to do that and re-map to something else completely -- but I doubt that this is happening.

Soo.... can you please explain what exactly are you trying to achieve by removing this `guardptr = mmap(0, total, PROT_NONE, flags, guardfd, 0);` logic?

Reply:
Instead of calling `mmap()` twice, we do `mmap(...PROT_READ|PROT_WRITE...)` + `mprotect(PROT_NONE)` to get the guard pages